### DISCUSSION TITLE
upsert where @@unique([id1, id2]) create missing relationship attributes as connect

### DISCUSSION THREAD

**Discussion by danwetherald** [prisma/prisma#5929] [Q&A]
Looks like when you use a upsert with a where that uses a `@@unique([])` the create will not allow attaching connections for belongsTo and hasMany.....

```prisma
model Trigger {
  id        String    @id @default(cuid())
  funnel    Funnel    @relation(fields: [funnelId], references: [id])
  funnelId  String
  product   Product   @relation(fields: [productId], references: [id])
  productId String
  variants  Variant[]

  @@unique([funnelId, productId])
}
```

```ts
return await context.prisma.trigger.upsert({
  where: {
    funnelId_productId: {
      funnelId: localFunnelId,
      productId: localProductId,
    },
  },
  create: {
    funnel: { connect: { id: localFunnelId } },
    product: { connect: { id: localProductId } },
    variants: localVariantIds.map((v: string) => ({
      connect: { id: v },
    })),
  },
  update: {
    variants: {
      set: localVariantIds.map((v: string) => ({ id: v })),
    },
  }
})

export type TriggerUncheckedCreateInput = {
  id?: string
  createdAt?: Date | string
  updatedAt?: Date | string
  deletedAt?: Date | string | null
  funnelId: string
  productId: string
}
```

```
PrismaClientValidationError: Unknown arg `funnel` in create.funnel for type TriggerUncheckedCreateInput. Did you mean `funnelId`?
Unknown arg `product` in create.product for type TriggerUncheckedCreateInput. Did you mean `productId`?
Unknown arg `variants` in create.variants for type TriggerUncheckedCreateInput.
Argument funnelId for create.funnelId is missing.
Argument productId for create.productId is missing.
```

I want to be able to connect the product and the variants in the create but there is only funnelId and productId - WHY?

`update: {}` works just fine, this renders upsert useless.

This also compiles with no errors or warning, looks like the XOR is not choosing the correct type?

`create: XOR<TriggerUncheckedCreateInput, TriggerCreateInput>` - Why is `TriggerCreateInput` not being used?

I have now even explicitly typed the create input as so and still same error, this is SO frustrating.

```ts
create: {
  funnel: { connect: { id: localFunnelId } },
  product: { connect: { id: localProductId } },
  variants: localVariantIds.map((v: string) => ({
    connect: { id: v },
  })),
} as Prisma.TriggerCreateInput,
```

**Comment by danwetherald** [comment]
This is the only solution aka HACK:

```ts
const trigger = await context.prisma.trigger.upsert({
  where: {
    funnelId_productId: {
      funnelId: localFunnelId,
      productId: localProductId,
    },
  },
  update: {
    variants: {
      set: localVariantIds.map((v: string) => ({ id: v })),
    },
  },
  create: {
    funnelId: localFunnelId,
    productId: localProductId,
  },
  include: {
    product: true,
    variants: true,
  },
});

// MORE PRISMA HACKS //
return await context.prisma.trigger.update({
  where: { id: trigger.id },
  data: {
    variants: {
      connect: localVariantIds.map((v: string) => ({ id: v })),
    },
  },
});
```

**Reply by ryands17** [reply]
@dan003400 

You're query was incorrect as you were mapping over connect. It should map over values:

```ts
  prisma.trigger.upsert({
    where: {
      funnelId_productId: {
        funnelId: localFunnelId,
        productId: localProductId,
      },
    },
    create: {
      funnel: { connect: { id: localFunnelId } },
      product: { connect: { id: localProductId } },
      variants: {
        connect: localVariantIds.map((v) => ({
          id: v,
        })),
      },
    },
    update: {
      variants: {
        set: localVariantIds.map((v) => ({ id: v })),
      },
    },
  })
```

`connect` accepts an [array of items](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#update-an-existing-user-record-by-connecting-it-to-two-existing-post-records) as mentioned in the docs. Add this and your query will work fine.

**Reply by danwetherald** [reply]
Gotcha, why is there no error in the typescript compiler then? Why is this error accepted?

**Reply by ryands17** [reply]
There actually should be an error shown in the compiler and your editor as well. VSCode shows me a red line with the error on hover. Which editor/IDE are you using currently?

**Reply by danwetherald** [reply]
There was no errors in the compiler, my IDE will get its errors from the compiler.

My editor of choice is Neovim.

**Reply by ryands17** [reply]
@dan003400 👋 

I think what you would need is [coc-tsserver](https://github.com/neoclide/coc-tsserver) as this comes from TypeScript and not Prisma. Also I would suggest installing [ale](https://github.com/dense-analysis/ale) for in-editor error and warnings.

**Reply by danwetherald** [reply]
I would tend to agree, but there was zero errors in typescript when this happen, I actually believe this might be more of a typescript issue with the prisma types than an editor problem.

Why is typescript allowing this incorrect data structure to be passed to the prisma client?

**Reply by ryands17** [reply]
That's not possible actually. This is my setup with the same above query. I am using VSCode:

![image](https://user-images.githubusercontent.com/19697099/109983270-abd5a000-7d28-11eb-965d-1b380f32261d.png)

And when I hover over it, I get this:

![image](https://user-images.githubusercontent.com/19697099/109983398-c9a30500-7d28-11eb-80fe-afe16e7e4c53.png)


So TypeScript should actually inform about this error directly in the editor with `tsserver`.

**Reply by ryands17** [reply]
The same setup in Vim with `coc-tsserver`:

![image](https://user-images.githubusercontent.com/19697099/109983998-606fc180-7d29-11eb-8454-2cd006ffb41e.png)

**Reply by danwetherald** [reply]
Odd, I have coc-tsserver installed, but not getting that error, any idea if this could be a configuration issue with the tsconfig?

**Reply by ryands17** [reply]
This is what I have in `tsconfig.json`:

```json
{
  "compilerOptions": {
    "sourceMap": true,
    "module": "commonjs",
    "target": "ES2018",
    "outDir": "dist",
    "lib": [
      "esnext",
    ],
    "strict": true,
    "strictNullChecks": true,
    "noImplicitAny": false,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "strictBindCallApply": true,
    "strictFunctionTypes": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "noUnusedParameters": false
  },
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
